File: erlang/etudes/ch6/non_fp.erl
-module(non_fp).
-export([generate_teeth/2]).

generate_teeth([],_) -> [];
generate_teeth([$F|Tail], Prob) -> [[0] | generate_teeth(Tail, Prob)];
generate_teeth([$T|Tail], Prob) ->
    random:seed(now()),
    [generate_tooth(Prob) | generate_teeth(Tail, Prob)].

generate_tooth(Prob) ->
    case random:uniform() < Prob of
        true -> generate_tooth(2, 6);
        false -> generate_tooth(3, 6)
    end.

generate_tooth(_, 0) -> [];
generate_tooth(Base, Count) -> 
    Offset = random:uniform(3) - 2,
    [Base + Offset | generate_tooth(Base, Count - 1)].

File: elixir/etudes/ch06/non_fp.ex
defmodule NonFP do
  def generate_pockets([],_), do: []
  def generate_pockets([?F|tail], prob), do: [[0] | generate_pockets(tail, prob)];
  def generate_pockets([?T|tail], prob) do
    :random.seed(:erlang.now())
    [generate_tooth(prob) | generate_pockets(tail, prob)]
  end

  defp generate_tooth(prob) do
    base_depth = if :random.uniform() < prob, do: 2, else: 3
    generate_tooth(base_depth, 6)
  end

  defp generate_tooth(_, 0), do: []
  defp generate_tooth(base, count) do
    offset = :random.uniform(3) - 2
    [base + offset | generate_tooth(base, count - 1)]
  end
end
